### ISSUE TITLE
Malformed version number string 9bb9d91 (ArgumentError)

### ISSUE THREAD

**Issue by Kieranflock** [pact-foundation/pact-python#201]
version:
pact-python (v 1.3.0)

use case:
Using pact-python to publish a consumer test to a pactflow broker. 

Reading the docs, it suggests using the commit SHA or something unique for the versioning. When passing other than a string in the format "x.x.x" I get the below error.

_example_
```
pact = Consumer(
    "example",
    tag_with_git_branch=True,
    tags=[prod_tag_flag]
).has_pact_with(
    Provider("exampleName"),
    host_name=mock_host,
    port=mock_port,
    publish_to_broker=True,
    broker_base_url=pact_broker_url,
    broker_token=pact_token,
    version="1.0.0" <---- ideally this would be dynamic and unique.
)
```

_error_
```
Malformed version number string 9bb9d91 (ArgumentError)
```

Am I misunderstanding the use of the `version` property here?
Or, is this where in fact I should be passing a string of the commit SHA for publication?

I have posted this into the pact slack channel also.

**Comment by Kieranflock** [comment]
I have overcome this error by structuring the version string as:

```
version='1.0.0-' + get_git_revision_short_hash()
```

Tbf this is how the graph in this [document](https://docs.pact.io/getting_started/versioning_in_the_pact_broker/) displays it (which was my hint) but I couldn't find a reference to this anywhere. 

I am not sure this is an issue?

**Comment by mefellows** [comment]
Whilst the semantic versioning approach linked to is probably recommended, we shouldn't be so strict on that. In fact, it's very common to simply use the git sha alone.

I might be mistaken here, but whilst the documentation for that constructor property says it's the consumer version, the implementation seems to [maps to the pact specification version](https://github.com/pact-foundation/pact-python/blob/master/pact/consumer.py#L111-L113) (and see also https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/pact.py#L115-L116), which pretty much needs to be the default `2.0.0`.

I think you only need to specify the version when you run the publish command (because that's the only time it's relevant).

**Comment by bethesque** [comment]
Are you using Pactflow or a self hosted open source Pact Broker? There is no parsing or validation done on the version number any more, and there hasn't been for many years. Please see https://docs.pact.io/pact_broker/configuration#ordering-versions

**Comment by Kieranflock** [comment]
Thanks both. 

I am using Pactflow for this @bethesque -- yes I didn't see any references in the docs for version number validation/structure aside from using a git SHA etc. 

Thanks @mefellows yes I did see that default value also but thought I could pass any string i.e. a SHA. If I am using `publish_to_broker=True,` to publish to the broker as part of pact-python and not say the Docker CLI then do I not need to be able to pass the consumer version into `version` as I did above?

Or are you saying there is another parameter for passing a consumer version number?

**Comment by mefellows** [comment]
TL;DR - the `version` field is used for both the specification version, and the application version and I believe is the cause of the bug.

So, I was slightly incorrect in my previous statement, but the problem remains: the version property passed to `has_pact_with` is used in two places with different meanings.

The main meaning appears to be the specification version, which is both documented and working as expected:

* The parameter is documented correctly for the specification case in the `Pact` class:: https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/pact.py#L115
* ...but not in the factory method in `consumer.py` which indicates its the application version: https://github.com/pact-foundation/pact-python/blob/0e9b71cd9a7ca1f543f678811cd86fc3c0b0b2f8/pact/consumer.py#L38
* It is assigned here https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/pact.py#L145
* It is passed to the CLI here: https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/pact.py#L199

For the publish case (when you enabled `publish_to_broker`, the field is re-used as the application version, which is incorrect:

* https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/pact.py#L238
* the documentation for it seems to be correct however: https://github.com/pact-foundation/pact-python/blob/64d7bdce2800dec5e1dacb7c013458f297b3593e/pact/consumer.py#L111

**Comment by elliottmurray** [comment]
@Kieranflock @mefellows #203 

Can you check that I have the right idea? If you have a working example to test it out that would be awesome as I don't! I have put some unit tests in but always good to get feedback.

**Comment by elliottmurray** [comment]
Should add I'll wait until I'm confident it actually works before I publish. So if you want to use clone and link locally.

**Comment by mefellows** [comment]
Hi @elliottmurray, apologies, I did see this but forgot to come back to it. It looks OK to my eye, but will need confirmation of somebody actually running the publish step this way. Thanks for fixing!

**Comment by elliottmurray** [comment]
The example e2e uses this flag. I'm not going to put a broken example in to prove it! I think the unit tests cover it.
